### PR TITLE
Add slow generate tests for pretrained lm models

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Move implementation of slow hardcoded generate models to this PR from. Checkout previous discussion in PR #2885

### CODE REVIEW

**Review by sshleifer — COMMENTED** [review]
Please incorporate my comments on this file in the other PR :)

**Comment by patrickvonplaten** [comment]
> Please incorporate my comments on this file in the other PR :)

is on my radar :-)

**Comment by patrickvonplaten** [comment]
UPDATE: Changed slow tests for language generation design according to discussion in PR #2885 .
If this looks alright, I'll add test cases for the other LMModels @LysandreJik & @sshleifer

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -587,12 +588,48 @@ def test_inputs_embeds(self):
...
+                        num_beams=config.num_beams,
+                    )
+
+                    self.assertListEqual(list(output_ids[0]), expected_output_ids)
+
 
 global_rng = random.Random()
 
 
+def list_to_torch_tensor(tokens_list):
```
Comment (sshleifer) [inline tests/test_modeling_common.py]: (nit) naming -> make_long_tensor?

File: tests/test_modeling_gpt2.py [hunk 2]
```diff
@@ -173,7 +175,7 @@ def create_and_check_lm_head_model(self, config, input_ids, input_mask, head_mas
 
             self.parent.assertListEqual(list(result["loss"].size()), [])
             self.parent.assertListEqual(
-                list(result["lm_logits"].size()), [self.batch_size, self.seq_length, self.vocab_size]
+                list(result["lm_logits"].size()), [self.batch_size, self.seq_length, self.vocab_size],
```
Comment (sshleifer) [inline tests/test_modeling_gpt2.py]: (nit) you can also do 
`self.parent.assertEqual(x.size(), (bs, seq_len, vocab_size)`
Which I would prefer but no strong feelings.

**Review by sshleifer — APPROVED** [review]
LGTM! I would get this one merged before you add more tests in order to keep the PRs easier to review.

**Comment by patrickvonplaten** [comment]
Updated the language model generation slow tests following Roberta's and Bart's Integration Test style. What do you think? @LysandreJik  @sshleifer

**Review by LysandreJik — COMMENTED** [review]
Glad we're working towards more complete integration tests!

I like that these tests are simple to understand and test the whole pipeline. Good for me to do this for all models that use generate.

**Review by sshleifer — APPROVED** [review]

File: src/transformers/modeling_xlm.py [hunk 3]
```diff
@@ -512,7 +512,7 @@ def forward(
             inputs_embeds = self.embeddings(input_ids)
 
         tensor = inputs_embeds + self.position_embeddings(position_ids).expand_as(inputs_embeds)
-        if langs is not None and self.use_lang_emb:
+        if langs is not None and self.use_lang_emb and self.n_langs > 1:
```
Comment (patrickvonplaten) [inline src/transformers/modeling_xlm.py]: I had to include this if statement to not get a `self.lang_embeddings` is undefined error. Is this fine @LysandreJik ?

**Comment by patrickvonplaten** [comment]
Finished to add hard-coded tests for all models with LMHead: `GPT2, OpenAI, XLNet, TransfoXL, CTRL and XLM.`

All pretrained models generate reasonable results **except** `XLM`. Might need to take a closer look in a future PR.

Also future PRs TODO:

- [ ] Add hardcoded tests for seq-to-seq language generation
- [ ] Add hardcoded tests for DoubleHead language generation

**Comment by sshleifer** [comment]
Don't feel strongly, but I would consider deleting the XLM example so that we don't ensure bad generations.
Also, if it is possible to make shorter examples (or more than one token per line), it would make the code more readable.
Overall, this is awesome!

**Comment by sshleifer** [comment]
We could also add a sentence to `Contributing.md` telling people who change the generation code which command to run to make sure they didn't break stuff!

File: src/transformers/modeling_xlm.py (hunk 3 above)
Comment (LysandreJik) [reply src/transformers/modeling_xlm.py]: I believe this is the exact same I did for [FlauBERT](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_flaubert.py#L234), too bad I missed XLM!

**Review by LysandreJik — APPROVED** [review]
This is great!

**Pull request merged**